docs/proposal/20201103-failure-domain.md

```diff
@@ -172,138 +173,207 @@ CSI/CPI will be configured with region/zone information so nodes and PVs will ha
-    DeploymentZones []string `json:"deploymentZones,omitempty"`
+  Spec   VSphereDeploymentZoneSpec   `json:"spec,omitempty"`
+  Status VSphereDeploymentZoneStatus `json:"status,omitempty"`
 }
-
 ```
 
 ### Implementation Details
 
-1. Manually Pre-populate CR for VSphereFailureDomain and VSphereDeploymentZone
-2. Controller change for user story 2/3/4
-3. Controller change for user story 1
-4. New controller for CRD VSphereFailureDomain
-5. Auto discovery, validation of VSphereFailureDomain for user story 1/2/3.
-6. Auto configuration for hostgroup for user story 1/2/3, if provided with enough permission
+the changes required to the controllers are the following
+
+#### VSphereFailureDomain validating webhook
+
+the validation for `VSphereFailureDomain` should:
+
+* verify that the `.Spec.Topology.Datacenter` is not empty
+* if `.Spec.Topology.HostGroup` is not `nil`, verify that `autoconfigure` is not true for both the `hostGroup` and `FailureDomain` struct
+
+#### VSphereDeploymentZone defaulting webhook
+
+the defaultijg for `VSphereDeploymentZone` should:
+
+* check if `controlPlane` is `nil`, if it is default to true
+
+#### the vspheredeploymentzone_controller
+
+This controller would be responsible for:
+
+* Listing `VSphereClusters` and checking if `server` field matches
+  * pickup the first matching `VSphereCluster` and extract credentials through reading `.spec.identityRef`
+  * No `VSphereCluster` is matching, this means that we fallback to the `capv-controller-manager` credentials
+* Verifying the following:
+  * being able to create a session
+  * the compute cluster exists and has the specified resource pool
+  * the network, datastore and folder all exist
+  * if autoconfigure is `false` verify that the hostGroup exists
+  * if autoconfigure is `false` verify that tags on the elements (compute cluster, datacenter or Hosts) exist
+* Depending on autoconfiguration enablement (only ONE of the following can be done):
+  * Create the hostGroup and add the tagged hosts
+  * List the hosts within the hostGroup and tag them accordingly
+* set `.status.Ready` when all of the above is done
+
+note: `.status.Ready` should remain nil, unless we deem the `VSphereDeploymentZone` not ready for consumption
+
+#### the vspherecluster_controller
+
+the following changes are going to be introduced to vspherecluster_controller:
+
+* List the `VSphereDeploymentZones` and match based on `.spec.server`
+* check `.status.Ready`
+  * if `.status.Ready` is `nil`, add it to the failureDomain map, but  skip setting `.status.InfrastructureReady` on the `VSphereCluster`
+  * if `.status.Ready` is `true` add it to the failureDomain map
+  * if `.status.Ready` is `false` skip adding `VSphereDeploymentZone` to the map
+* copy `.spec.controlPlane` of `VSphereDeploymentZone` into the failureDomain map value
+* Before setting `.status.InfrastructureReady` ensure that no matched `VSphereFailureDomain` has `.status.Ready` set to `nil`
```

Comment:
What about the case, if any of them moves from being `nil` to `false`. Should we remove the FD from the map and then add the condition about skipping it.